add timeout option for flushing stream before highWaterMark

rjferguson21 opened the pull request:
We had a use case where a stream potentially would never end.  This resulted in a situation where it was desirable flush the queue after a configurable timeout to avoid the remaining values being stuck waiting for the stream to end or for the next operation where more records would flush them out.

rjferguson21 committed:
Add timeout to flush stream

rjferguson21 committed:
add tests for timeout

voldern commented:
Hi,

Sorry for the slow reply and thank you so much for the pull request!

I'll get this merged tomorrow. :+1:

Commit:
Rename timeout to flushTimeout

Commit:
Document flushTimeout option

Commit:
Use `.bind()` instead of `self = this`

Commit:
Update test description to fit naming pattern

Commit:
Remove superflous test and use sinon timers

Commit:
Rename timeoutId to flushTimeoutId

Commit:
Clear timeout when highWaterMark has been reached

Commit:
Document flushTimeout in README

Commit:
Remove superflous function in test

voldern commented:
I've added a pull request back into your branch with some proposed changes: https://github.com/rjferguson21/elasticsearch-bulk-index-stream/pull/1.

Not sure if its the best way to go about this, but its the only one I could think of.

Commit:
Clear queue before giving up control of thread

Possible race condition pointed out by rjferguson21:
https://github.com/rjferguson21/elasticsearch-bulk-index-stream/pull/1#issuecomment-193408031.

Clear queue before giving up control of thread.

rjferguson21 committed:
Merge pull request #1 from voldern/timeout

Code review

voldern commented:
v0.2.0 with flushTimeout has been released to NPM. Thanks again @rjferguson21 :bow:
